internal/trace/transport.go

```diff
@@ -53,35 +54,42 @@ func (t *Transport) RoundTrip(req *http.Request) (resp *http.Response, err error
 	ctx := req.Context()
 	e := log.GetLogger(ctx)
 
-	e.Debugf("> Request: %q %q", req.Method, req.URL)
-	e.Debugf("> Request headers:")
-	logHeader(req.Header, e)
+	// logs to be printed out
+	var logs strings.Builder
+	fmt.Fprintf(&logs, "> Request: %q %q\n", req.Method, req.URL)
+	fmt.Fprintln(&logs, "> Request headers:")
+	fmt.Fprint(&logs, logHeader(req.Header))
 
 	resp, err = t.RoundTripper.RoundTrip(req)
 	if err != nil {
+		e.Debug(logs.String())
 		e.Errorf("Error in getting response: %w", err)
 	} else if resp == nil {
+		e.Debug(logs.String())
 		e.Errorf("No response obtained for request %s %q", req.Method, req.URL)
 	} else {
-		e.Debugf("< Response status: %q", resp.Status)
-		e.Debugf("< Response headers:")
-		logHeader(resp.Header, e)
+		fmt.Fprintf(&logs, "< Response status: %q\n", resp.Status)
+		fmt.Fprintln(&logs, "< Response headers:")
+		fmt.Fprint(&logs, logHeader(resp.Header))
+		e.Debug(logs.String())
 	}
 	return resp, err
 }
 
-// logHeader prints out the provided header keys and values, with auth header
+// logHeader returns string of provided header keys and values, with auth header
 // scrubbed.
-func logHeader(header http.Header, e log.Logger) {
+func logHeader(header http.Header) string {
```

Comment:
```suggestion
func logHeader(header http.Header, logs strings.Builder) string {
```

Is it better to write to builder directly to reduce a copy operation.

Reply:
The above suggestion would panic due to `non-zero Builder copied by value`: https://pkg.go.dev/strings#Builder

Passing the pointer is an alternative here.

Reply:
Yes, we need a pointer for `strings.Builder`. I want to improve performance by avoiding unnecessary memory allocation and copying. In line 73, we allocate memory and copy the return from `logHeader`. This operation can be eliminated if we directly write to the `logs`.